pkg/frontend/v2/frontend_scheduler_adapter.go

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package v2
+
+import (
+	"context"
+	"time"
+
+	"github.com/grafana/dskit/httpgrpc"
+	"github.com/grafana/dskit/tenant"
+
+	"github.com/grafana/mimir/pkg/frontend/querymiddleware"
+	v1 "github.com/grafana/mimir/pkg/frontend/v1"
+	"github.com/grafana/mimir/pkg/querier"
+	"github.com/grafana/mimir/pkg/scheduler/schedulerpb"
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+type frontendToSchedulerAdapter struct {
+	cfg             Config
+	limits          v1.Limits
+	prometheusCodec querymiddleware.Codec
+}
+
+func (a *frontendToSchedulerAdapter) frontendToSchedulerEnqueueRequest(
+	req *frontendRequest, frontendAddr string,
+) *schedulerpb.FrontendToScheduler {
+	return &schedulerpb.FrontendToScheduler{
+		Type:                      schedulerpb.ENQUEUE,
+		QueryID:                   req.queryID,
+		UserID:                    req.userID,
+		HttpRequest:               req.request,
+		FrontendAddress:           frontendAddr,
+		StatsEnabled:              req.statsEnabled,
+		AdditionalQueueDimensions: a.extractAdditionalQueueDimensions(req.ctx, req.request),
+	}
+}
+
+const ShouldQueryIngestersQueueDimension = "ingester"
+const ShouldQueryStoreGatewayQueueDimension = "store-gateway"
+const ShouldQueryIngestersAndStoreGatewayQueueDimension = "ingester|store-gateway"
```

Comment:
open to suggestions what exactly these constants are.
In terms of the functioning of the program, their contents do not matter, they could be "a", "b", "c".

However we expect we may want to add some observability on the separate queue lengths
We currently don't add any queue dimension labels to the `cortex_query_scheduler_queue_duration_seconds` histogram, but if we want to add observability on the queue duration for these three types of queues, we would probably want to apply these labels unmodified, or map them to a more prometheus-friendly representation.